sklearn/manifold/_t_sne.py

```diff
@@ -959,7 +961,14 @@ def _fit(self, X, skip_num_points=0):
                 )
 
             t0 = time()
-            distances_nn = knn.kneighbors_graph(mode="distance")
+            with config_context(enable_cython_pairwise_dist=False):
+                # Distances have really tiny difference if using the
+                # PairwiseDistancesReductions on 32bit data.
+                # This causes in turn the joint probability distributions computations.
+                # below to have significant differences and thus different results.
+                # For backward compatibility on results, we prefer to use the
+                # old back-end.
```

Comment:
I don't understand this comment. If the problem is about `float32` vs `float64`, we could enable the old numpy-based backend only for `float32` data and still use the `float64` for the new backend, no? Otherwise we will introduce a performance regression for `float64` data in 1.2 with respect to 1.1.

Do you understand why would the old backend be more numerically accurate than the new backend?

Which T-SNE tests in particular would fail if we were to use the new backend here?

Reply:
Yes, this is a really coarse restriction that I will refine.

---

If we remove the context, this test fails:
https://github.com/scikit-learn/scikit-learn/blob/d2f6a66c1740abe477d5bdffed62aa6e8e345fc6/sklearn/manifold/tests/test_t_sne.py#L1110-L1135

It does not fail on `main` because `sklearn.metrics.manhattan_distances` is used on both variant under the hood.

On this PR, distances get computed via `ManhattanDistance` in the `PairwiseDistancesReduction` infrastructure and there are small differences when using float32 for the default `method`.

Namely, when `method="barnes-hut"` for TNSE, `kneighbors_graph` which ultimately uses `PairwiseDistancesArgkmin` via `kneighbors` is called:

https://github.com/scikit-learn/scikit-learn/blob/f9f3fee387fda933c3612c893f9ff3e213b460be/sklearn/manifold/_t_sne.py#L933-L962

the max of the abs of the element-wise difference is `10e-7` for the distance matrix (`A_data`): 

https://github.com/scikit-learn/scikit-learn/blob/bcefbe1e3a5bc5ff112e4b2caf748e0d1fa6a304/sklearn/neighbors/_base.py#L989

Those small numerical differences are then propagated in those computations:
https://github.com/scikit-learn/scikit-learn/blob/f9f3fee387fda933c3612c893f9ff3e213b460be/sklearn/manifold/_t_sne.py#L974-L982

causing `P` to be slight different (`max(abs(diff_P)) = 10e-10`) and the gradient descent to give different results (`max(abs(diff_params)) = 18.360132` in this case):

https://github.com/scikit-learn/scikit-learn/blob/f9f3fee387fda933c3612c893f9ff3e213b460be/sklearn/manifold/_t_sne.py#L1067-L1075

---

Reply:
Ok thanks for the investigation. So those the root cause are expected machine precision level rounding errors.

This problem would disappear if we changed `manhattan_distances` to also use the new infrastructure, right?

Let's change this test to explain this and call `pytest.xfail`. I think in a follow-up PR we might want to refactor `sklearn.metrics.manhattan_distances` to rely on the new code which will make the problem go away, but let's do one thing at a time.

Reply:
> This problem would disappear if we changed manhattan_distances to also use the new infrastructure, right?

Yes. More generally, I think it would help to clarify the reliance on `scipy.spatial.distance` for scikit-learn.

Reply:
@thomasjpfan: FYI, here is a follow-up of discussions regarding failure for t-SNE.